src/badge/withBadge.js

```diff
@@ -0,0 +1,57 @@
+import React from 'react';
+import { StyleSheet, View } from 'react-native';
+
+import Badge from './Badge';
+
+const withBadge = (value, options = {}) => WrappedComponent => {
+  class WithBadge extends React.Component {
+    render() {
+      const {
+        left = -7,
```

Comment:
What are these values based on?

![image](https://user-images.githubusercontent.com/5962998/49702595-396cc880-fbd0-11e8-9ee2-40c5f2cda96e.jpeg)

```jsx
const BadgedAvatar = withBadge(1, {
    status: 'success',
  })(Avatar);

  const BadgedIcon = withBadge(1, {
    status: 'success',
  })(Icon);

<BadgedAvatar size="medium" title="MD" rounded />
<BadgedIcon name="chat" size={30} />
```

Any ideas for defaults?

Reply:
It's to the left by default I see

Reply:
They are based on icons for headers and tabs in react navigation.

```js
export class SomeScreen extends Component {
  static navigationOptions = () => {
    const BadgedIcon = withBadge(1)(Icon);
    return {
      headerTitle: "Title",
      headerRight: (
        <BadgedIcon
          name={`${Platform.OS === "ios" ? "ios" : "md"}-chatbubbles`}
          type="ionicon"
          color="white"
          containerStyle={styles.padRight}
        />
      )
    };
  };
```

But, since this HOC lives here, it might be better to base them on components for RNE 😄 So should we optimize the defaults for `<Avatar />`s in `<ListItem />`s?

```js
renderItem = ({ item }) => {
  const BadgedAvatar = withBadge(1)(Avatar);
  return (
    <ListItem
      title={item.name}
      leftAvatar={<BadgedAvatar source={{ uri: item.avatar }} rounded={true} />}
      bottomDivider={true}
    />
  );
};
```

Reply:
We can decide now. The badge defaults are pretty big, so i think i'll change the sizes of the original badge. I took some ideas from the twitter app.

![screen shot 2018-12-09 at 5 46 25 pm](https://user-images.githubusercontent.com/5962998/49703362-763dbd00-fbda-11e8-9600-bd3539185b30.jpg)

Will make those changes then let you know

Reply:
Okay. I took the sizes (that I corrected earlier today), from iOS (WhatsApp and Telegram).

Reply:
Alright updated! It looks pretty good.

Should we assume the badge will probably go to the top right?

Reply:
![screenshot_1544398184](https://user-images.githubusercontent.com/5962998/49704401-d6d3f680-fbe8-11e8-8a39-3e86cf16c5bf.png)

Whatcha think? This is with:
```js
const {
  left = 45,
  bottom = null,
  top = 4,
  right = null,
  hidden = false,
  ...badgeProps
} = options;
```

cc @janhesters @xavier-villelegier

Reply:
I think it should be a little more to the right, and a little more down. Here is an example from WhatsApp to compare:

![whatsappexample](https://user-images.githubusercontent.com/31096420/49717065-184ebb00-fc56-11e8-8daf-b7efb5ff14e7.jpg)

Also in the right most tab for status we have a reference for the mini badge. Furthermore if the value of the badge is in the double digits, as in the second listitem in this screen shot, it should have flat top and bottom borders (I also tested this behavior in WhatsApp with the badge from the tabs by having a friend spamming me 😄). Btw this look is consistent in Telegram, too.

Reply:
Okay, I tweaked it to resemble the WhatsApp look:

![withbadgeexample](https://user-images.githubusercontent.com/31096420/49722736-b433f300-fc65-11e8-8e78-3c070192d962.png)

Here is the change that I did for that:

```js
const { bottom, hidden = false, left, ...badgeProps } = options;a
let { right = -16, top = -1 } = options;
if (!value) {
  (right = -3), (top = 3);
}
```

@iRoachie Either way, when default the badge to the top right, we should use top and right as reference values instead of left (or bottom).

**On another note:** Notice how in your Twitter the big badge example and in my WhatsApp example the mini badge, has a background around itself. Maybe we should add a boolean to the `<Badge />` that does this? 🤔 Otherwise people would have to do this styling manually.

Reply:
The mini badge doesn't have any additional styling other than size https://github.com/react-native-training/react-native-elements/blob/next/src/badge/Badge.js#L72. Seems to be a bug with border radius and background color, but it only happens on iOS.

```jsx
if (!value) {
  (right = -3), (top = 3);
}
```

^ I think this might a bit too much logic.

> @iRoachie Either way, when default the badge to the top right, we should use top and right as reference values instead of left (or bottom).


The right I did the value from the left is so that when the badge value it increases that it doesn't move and cover the icon. e.g `99+` would start to cover the icon. If the positioning is set from the left then it never covers.

> Furthermore if the value of the badge is in the double digits, as in the second listitem in this screen shot, it should have flat top and bottom borders 

The badge does this already!

Reply:
>The badge does this already!

Just noticed this, too (awesome).

>^ I think this might a bit too much logic.

Sometimes extra logic is good. I mean without this line, the MiniBadge would always be off, resulting in each user having to add their own position. I think the extra logic is worth taking of this labour step :)

BTW, I see another problem, that both our suggestions have:

A very common thing for (icon-/ text) button's in the header is to add padding or margin so that they don't stick to the side of the screen.

E.g. using React Navigation and RNE:

```js
  public static navigationOptions = () => {
    const BadgedIcon = withBadge()(Icon);
    return {
      headerLeft: <Button title="Edit" type="clear" containerStyle={styles.padLeft} />,
      headerRight: (
        <BadgedIcon name="edit" type="feather" containerStyle={styles.padRight} color="#007aff" />
      ),
      title: 'Chats'
    };
  };
```

Now the problem with this is that with the current way this HOC is written it will expand the container `<View />` of the icon and place the `<Badge />` related to this container. This results in this:

![headerbug](https://user-images.githubusercontent.com/31096420/49731188-d4ba7800-fc7a-11e8-928a-d7fd53b6c8c0.png)

Any idea for a solution to tackle all these little details?

Reply:
Okay actually the last thing might not be that big of a problem, cause one could always do:

```js
  public static navigationOptions = () => {
    const BadgedIcon = withBadge()(Icon);
    return {
      headerLeft: <Button title="Edit" type="clear" containerStyle={styles.padLeft} />,
      headerRight: (
        <View style={styles.padRight}>
          <BadgedIcon name="edit" type="feather" color="#007aff" />
        </View>
      ),
      title: 'Chats'
    };
  };
```

Or we add a parameter to the HOC that let's the user define padding for the HOC's container?

Reply:
> A very common thing for (icon-/ text) buttons within the header is to add padding or margin so that they don't stick to the side of the screen.

I understand the needed for it but i'm not sure why a dev would use padding here and not margin. Wouldn't margin be more fitting here?

> Or we add a parameter to the HOC that let's the user define styles for the HOC's container?
I agree with this also!

Reply:
Done. The only thing that we have to think about now are the default values.